lang/py/src/avro/io.py

```diff
@@ -480,11 +633,32 @@ def read_data(self, writers_schema, readers_schema, decoder):
     elif writers_schema.type == 'boolean':
       return decoder.read_boolean()
     elif writers_schema.type == 'string':
-      return decoder.read_utf8()
+      if (hasattr(writers_schema, 'logical_type') and
+          writers_schema.logical_type == constants.DECIMAL):
+        return decoder.read_decimal_from_string()
+      else:
+        return decoder.read_utf8()
     elif writers_schema.type == 'int':
-      return decoder.read_int()
+      if (hasattr(writers_schema, 'logical_type') and
+          writers_schema.logical_type == constants.DATE):
+        return decoder.read_date_from_int()
+      elif (hasattr(writers_schema, 'logical_type') and
+        writers_schema.logical_type == constants.TIME_MILLIS):
+        return decoder.read_time_millis_from_int()
+      else:
+        return decoder.read_int()
     elif writers_schema.type == 'long':
-      return decoder.read_long()
+      if (hasattr(writers_schema, 'logical_type') and
+          writers_schema.logical_type == constants.TIME_MICROS):
+        return decoder.read_time_micros_from_long()
+      elif (hasattr(writers_schema, 'logical_type') and
+            writers_schema.logical_type == constants.TIMESTAMP_MILLIS):
+        return decoder.read_timestamp_millis_from_long()
+      elif (hasattr(writers_schema, 'logical_type') and
```

Comment:
I see what you mean about duplication of code here. It's annoying..

Reply:
makes me want to shake the original author: "put all your complicated logic in one place!!"